src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -300,125 +301,73 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
-        // We need to modify the reference there as well to forge
-        // Divide(sum/count) + 1
-        case aggExpr @ AggregateExpression(Average(ref), _, _, _) =>
-          // Need a type promotion
-          val sumToPush = newAggregate(Sum(ref), aggExpr)
-          val countToPush = newAggregate(Count(ref), aggExpr)
-
-          // Need a new expression id since they are not simply rewrite as above
-          val sumFinal = newAggregateWithId(Sum(toAlias(sumToPush).toAttribute), aggExpr)
-          val countFinal = newAggregateWithId(Sum(toAlias(countToPush).toAttribute), aggExpr)
-
-          avgPushdownRewriteMap(aggExpr.resultId) = List(sumToPush, countToPush)
-          avgFinalRewriteMap(aggExpr.resultId) = List(sumFinal, countFinal)
-          List(sumFinal, countFinal)
-        case _ => aggExpr :: Nil
+        case e: Max     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: Min     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: Sum     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: First   => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case _: Count   => aggExpr.copy(aggregateFunction = Sum(partialResultRef))
+        case _: Average => throw new IllegalStateException("All AVGs should have been rewritten.")
+        case _          => aggExpr
       }
     }
 
-    val pushdownAggregates = aggregateExpressions.flatMap { aggExpr =>
-      avgPushdownRewriteMap
-        .getOrElse(aggExpr.resultId, List(aggExpr))
-    }.distinct
-
-    aggregationToDAGRequest(groupingExpressions, pushdownAggregates, source, dagReq)
-
-    val rewrittenResultExpression = resultExpressions.map(
-      expr =>
-        expr
-          .transformDown {
-            case aggExpr: AttributeReference if avgFinalRewriteMap.contains(aggExpr.exprId) =>
-              // Replace the original Average expression with Div of Alias
-              val sumCountPair = avgFinalRewriteMap(aggExpr.exprId)
-
-              // We missed the chance for auto-coerce already
-              // so manual cast needed
-              // Also, convert into resultAttribute since
-              // they are created by tiSpark without Spark conversion
-              // TODO: Is DoubleType a best target type for all?
-              Cast(
-                Divide(
-                  Cast(sumCountPair.head.resultAttribute, DoubleType),
-                  Cast(sumCountPair(1).resultAttribute, DoubleType)
-                ),
-                aggExpr.dataType
-              )
-            case other => other
-          }
-          .asInstanceOf[NamedExpression]
-    )
+    aggregationToDAGRequest(groupingExpressions, aggregateExpressions.distinct, source, dagReq)
+
+    projects
+      .map { _.toAttribute.name }
+      .map { TiColumnRef.create }
+      .foreach { dagReq.addRequiredColumn }
```

Comment:
I see, missed the `addRequiredColumn` call while cleaning up the original code path, thanks for fixing it!

Reply:
Not your problem, this code snippet was just introduced into `master` branch few hours ago via this PR https://github.com/pingcap/tispark/pull/143, and I merged that change into this branch.